### PR TITLE
Fix several outstanding small formatter bugs

### LINKED ISSUE #31571: Incorrect formatting of conversion operator to nullable type
**Version Used**: Visual Studio 2019 Preview 1

**Steps to Reproduce**:

Run Format Document on the following code:

```csharp
public class MyType
{
    public static explicit operator int?(MyType obj) => null;
}
```

**Expected Behavior**:

```csharp
public class MyType
{
    public static explicit operator int?(MyType obj) => null;
}
```

**Actual Behavior**:

```csharp
public class MyType
{
    public static explicit operator int? (MyType obj) => null;
}
```

### LINKED ISSUE #33458: C# formatter missing linebreak after curly braces of for block
When using an for block inside an if condition without curly braces, the else gets put in the same line as the closing brace of the for block.

example:

```csharp
if (true)
    for (int i = 0; i <2; i++) {
       i++;
    } else
     b =5;
```

instead of the expected:

```csharp
if (true)
    for (int i = 0; i <2; i++) {
       i++;
    } 
else
     b =5;
```

correct placement:

```csharp
if(true){
   b=2;
} else
   b=3;
```

and:

```csharp
if(true)
   b=2;
else
   b=4;
```

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/431619/c-formatter-missing-linebreak-after-curly-braces-o.html
VSTS ticketId: 774047_
_These are the original issue comments:_

roland.fredenhagen on 1/20/2019, 06:42 AM (27 days ago): The Title misses "else" sorry for that, but it seams like I am not able to edit anything.
_These are the original issue solutions:_
(no solutions)

### LINKED ISSUE #33839: Intendation for switch expression is not good in expression bodied method
**Version Used**: 
VS 2019 RC
image
image
image
The first case should be considered mis-aligned, comparing to object initializer.

### LINKED ISSUE #33910: Incorrect formatter behavior in implicit conversions to a tuple
```C#
class C
{
    public static implicit operator (C, C)(C c) => throw null;
    public static implicit operator int?(C c) => throw null;
}
```

Currently, the formatter wants to insert a space between the type and the argument list here. It should not, like all other conversions.

### PR DESCRIPTION
**Pull request by sharwell**
📝 I can break this pull request up on request.

* `NewLineForElse` must check parent of preceding brace (fixes #33458)
* Fix spacing before parameter list for conversion operators (fixes #31571, fixes #33910)
* Fix formatting of switch expression blocks (fixes #33839)

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@sharwell can you let me know when you'd like a review?  Thanks!

**Comment by sharwell** [comment]
@CyrusNajmabadi this is now ready for review

File: src/Workspaces/Core/Portable/Shared/Extensions/ChildSyntaxListExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+using System;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Shared.Extensions
+{
+    internal static class ChildSyntaxListExtensions
+    {
+        public static SyntaxNodeOrToken First(this ChildSyntaxList childSyntaxList, Func<SyntaxNodeOrToken, bool> predicate)
+        {
+            foreach (var syntaxNodeOrToken in childSyntaxList)
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Shared/Extensions/ChildSyntaxListExtensions.cs]: why not just use enumerable? to reduce allocation?

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs [hunk 2]
```diff
@@ -9004,5 +9068,31 @@ public void Foo()
     }
 }");
         }
+
+        [Theory, Trait(Traits.Feature, Traits.Features.Formatting)]
+        [WorkItem(31571, "https://github.com/dotnet/roslyn/issues/31571")]
+        [WorkItem(33910, "https://github.com/dotnet/roslyn/issues/33910")]
+        [CombinatorialData]
```
Comment (heejaechang) [inline src/Workspaces/CSharpTest/Formatting/FormattingTests.cs]: oh. new stuff. what does this do?

File: src/Workspaces/CSharp/Portable/Formatting/Rules/IndentBlockFormattingRule.cs [hunk 3]
```diff
@@ -150,6 +152,14 @@ private void AddAlignmentBlockOperation(List<IndentBlockOperation> list, SyntaxN
                 case ImplicitArrayCreationExpressionSyntax implicitArrayCreation when implicitArrayCreation.Initializer != null:
                     SetAlignmentBlockOperation(list, implicitArrayCreation.NewKeyword, implicitArrayCreation.Initializer.OpenBraceToken, implicitArrayCreation.Initializer.CloseBraceToken, IndentBlockOption.RelativeToFirstTokenOnBaseTokenLine);
                     return;
+                case CSharpSyntaxNode syntaxNode when syntaxNode.IsKind(SyntaxKindEx.SwitchExpression):
+                    SetAlignmentBlockOperation(
+                        list,
+                        syntaxNode.GetFirstToken(),
+                        syntaxNode.ChildNodesAndTokens().First(child => child.IsKind(SyntaxKind.OpenBraceToken)).AsToken(),
```
Comment (heejaechang) [inline src/Workspaces/CSharp/Portable/Formatting/Rules/IndentBlockFormattingRule.cs]: can we just cast node to switchexpression and use open brace and close brance token directly ?

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/CSharpTest/Formatting/FormattingTests.cs]: Tests every combination of the inputs.

File: src/Workspaces/Core/Portable/Shared/Extensions/ChildSyntaxListExtensions.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/Extensions/ChildSyntaxListExtensions.cs]: > to reduce allocation?

Correct. The formatter is a huge source of allocations on a performance hot path.

File: src/Workspaces/CSharp/Portable/Formatting/Rules/IndentBlockFormattingRule.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Formatting/Rules/IndentBlockFormattingRule.cs]: No, `SwitchExpressionSyntax` does not exist in the Code Style layer.

**Pull request merged**